File: app/audit/analyzers/anomaly.py
"""
Anomaly Detection Analyzer
Implements statistical methods to detect unusual patterns in SEO metrics
"""

from typing import Dict, List, Optional
import numpy as np
from statistics import mean, stdev

from app.audit.models import (
    AuditIssue, SEOMetrics,
    IssueSeverity, IssueCategory, IssueType
)


class AnomalyDetector:
    """
    Detects anomalies in SEO metrics using statistical methods
    Based on research: Z-score, IQR, and threshold-based detection
    """
    
    def __init__(self):
        """Initialize anomaly detector with thresholds"""
        
        # Z-score thresholds for anomaly detection
        self.z_score_thresholds = {
            'warning': 2.0,    # 2 standard deviations (95% confidence)
            'critical': 3.0    # 3 standard deviations (99.7% confidence)
        }
        
        # Percentage change thresholds
        self.change_thresholds = {
            'traffic': {
                'critical': -50,  # >50% traffic loss
                'high': -20,      # 20-50% traffic loss
                'medium': -10     # 10-20% traffic loss
            },
            'position': {
                'critical': 5,    # Dropped 5+ positions
                'high': 3,        # Dropped 3-5 positions
                'medium': 2       # Dropped 2-3 positions
            },
            'ctr': {
                'critical': -50,  # >50% CTR drop
                'high': -30,      # 30-50% CTR drop
                'medium': -15     # 15-30% CTR drop
            },
            'impressions': {
                'critical': -60,  # >60% impression loss
                'high': -30,      # 30-60% impression loss
                'medium': -15     # 15-30% impression loss
            }
        }
    
    async def detect_anomalies(
        self,
        current_data: Dict,
        historical_data: Dict,
        metrics: SEOMetrics
    ) -> List[AuditIssue]:
        """
        Detect anomalies in SEO metrics
        
        Args:
            current_data: Current period metrics
            historical_data: Historical data with baselines
            metrics: Current SEO metrics
            
        Returns:
            List of detected anomaly issues
        """
        
        issues = []
        
        # 1. Traffic Drop Detection
        traffic_issue = self._detect_traffic_anomaly(
            current_data, historical_data, metrics
        )
        if traffic_issue:
            issues.append(traffic_issue)
        
        # 2. Position Loss Detection
        position_issue = self._detect_position_anomaly(
            current_data, historical_data, metrics
        )
        if position_issue:
            issues.append(position_issue)
        
        # 3. CTR Decline Detection
        ctr_issue = self._detect_ctr_anomaly(
            current_data, historical_data, metrics
        )
        if ctr_issue:
            issues.append(ctr_issue)
        
        # 4. Impression Drop Detection
        impression_issue = self._detect_impression_anomaly(
            current_data, historical_data, metrics
        )
        if impression_issue:
            issues.append(impression_issue)
        
        # 5. Statistical Anomaly Detection using Z-score
        statistical_issues = self._detect_statistical_anomalies(
            current_data, historical_data, metrics
        )
        issues.extend(statistical_issues)
        
        return issues
    
    def _detect_traffic_anomaly(
        self,
        current_data: Dict,
        historical_data: Dict,
        metrics: SEOMetrics
    ) -> Optional[AuditIssue]:
        """Detect significant traffic drops"""
        
        current_clicks = metrics.total_clicks
        previous_clicks = historical_data.get('previous_clicks', 0)
        
        if previous_clicks == 0:
            return None
        
        # Calculate percentage change
        change_pct = ((current_clicks - previous_clicks) / previous_clicks) * 100
        
        # Determine severity based on thresholds
        severity = None
        thresholds = self.change_thresholds['traffic']
        
        if change_pct <= thresholds['critical']:
            severity = IssueSeverity.CRITICAL
        elif change_pct <= thresholds['high']:
            severity = IssueSeverity.HIGH
        elif change_pct <= thresholds['medium']:
            severity = IssueSeverity.MEDIUM
        
        if severity:
            # Calculate z-score if baseline data available
            z_score = None
            if 'clicks_mean' in historical_data and 'clicks_stdev' in historical_data:
                if historical_data['clicks_stdev'] > 0:
                    z_score = (current_clicks - historical_data['clicks_mean']) / historical_data['clicks_stdev']
            
            issue = AuditIssue(
                issue_type=IssueType.TRAFFIC_DROP,
                severity=severity,
                category=IssueCategory.PERFORMANCE,
                title=f"Significant Traffic Drop Detected",
                description=f"Organic traffic has dropped by {abs(change_pct):.1f}% compared to the previous period. "
                           f"Current: {current_clicks} clicks, Previous: {previous_clicks} clicks.",
                recommendation="Investigate recent changes to your site, check for algorithm updates, "
                              "and review your top-performing pages for issues.",
                traffic_impact=abs(change_pct),
                business_impact="high" if severity == IssueSeverity.CRITICAL else "medium",
                evidence={
                    'current_clicks': current_clicks,
                    'previous_clicks': previous_clicks,
                    'change_percentage': change_pct,
                    'z_score': z_score
                }
            )
            
            return issue
        
        return None
    
    def _detect_position_anomaly(
        self,
        current_data: Dict,
        historical_data: Dict,
        metrics: SEOMetrics
    ) -> Optional[AuditIssue]:
        """Detect significant position losses"""
        
        current_position = metrics.average_position
        previous_position = historical_data.get('previous_position', 0)
        
        if previous_position == 0 or current_position == 0:
            return None
        
        # Position change (positive means dropped in rankings)
        position_change = current_position - previous_position
        
        # Determine severity based on thresholds
        severity = None
        thresholds = self.change_thresholds['position']
        
        if position_change >= thresholds['critical']:
            severity = IssueSeverity.CRITICAL
        elif position_change >= thresholds['high']:
            severity = IssueSeverity.HIGH
        elif position_change >= thresholds['medium']:
            severity = IssueSeverity.MEDIUM
        
        if severity:
            # Calculate statistical significance
            z_score = None
            if 'position_mean' in historical_data and 'position_stdev' in historical_data:
                if historical_data['position_stdev'] > 0:
                    z_score = (current_position - historical_data['position_mean']) / historical_data['position_stdev']
            
            issue = AuditIssue(
                issue_type=IssueType.POSITION_LOSS,
                severity=severity,
                category=IssueCategory.PERFORMANCE,
                title=f"Average Position Declined",
                description=f"Your average search position has dropped from {previous_position:.1f} to {current_position:.1f} "
                           f"(declined by {position_change:.1f} positions).",
                recommendation="Review your content quality, check for new competitors, "
                              "and ensure your pages are optimized for target keywords.",
                traffic_impact=position_change * 10,  # Estimate impact
                business_impact="high" if severity == IssueSeverity.CRITICAL else "medium",
                evidence={
                    'current_position': current_position,
                    'previous_position': previous_position,
                    'position_change': position_change,
                    'z_score': z_score
                }
            )
            
            return issue
        
        return None
    
    def _detect_ctr_anomaly(
        self,
        current_data: Dict,
        historical_data: Dict,
        metrics: SEOMetrics
    ) -> Optional[AuditIssue]:
        """Detect CTR problems despite impressions"""
        
        current_ctr = metrics.average_ctr
        previous_ctr = historical_data.get('previous_ctr', 0)
        current_impressions = metrics.total_impressions
        
        # Check for CTR decline with maintained/increased impressions
        if previous_ctr == 0:
            return None
        
        # Calculate CTR change
        ctr_change_pct = ((current_ctr - previous_ctr) / previous_ctr) * 100
        
        # Special case: CTR declined but impressions are stable/growing
        impressions_stable = current_impressions >= historical_data.get('previous_impressions', 0) * 0.9
        
        # Determine severity
        severity = None
        thresholds = self.change_thresholds['ctr']
        
        if ctr_change_pct <= thresholds['critical']:
            severity = IssueSeverity.CRITICAL if impressions_stable else IssueSeverity.HIGH
        elif ctr_change_pct <= thresholds['high']:
            severity = IssueSeverity.HIGH if impressions_stable else IssueSeverity.MEDIUM
        elif ctr_change_pct <= thresholds['medium'] and impressions_stable:
            severity = IssueSeverity.MEDIUM
        
        if severity:
            issue = AuditIssue(
                issue_type=IssueType.CTR_DECLINE,
                severity=severity,
                category=IssueCategory.PERFORMANCE,
                title="Click-Through Rate Declined" + (" Despite Stable Impressions" if impressions_stable else ""),
                description=f"CTR has dropped from {previous_ctr:.2f}% to {current_ctr:.2f}% "
                           f"({abs(ctr_change_pct):.1f}% decline)." +
                           (" Your pages are showing in search but not attracting clicks." if impressions_stable else ""),
                recommendation="Review and improve your title tags and meta descriptions. "
                              "Ensure they are compelling and match search intent.",
                traffic_impact=abs(ctr_change_pct),
                business_impact="high" if impressions_stable else "medium",
                evidence={
                    'current_ctr': current_ctr,
                    'previous_ctr': previous_ctr,
                    'ctr_change': ctr_change_pct,
                    'impressions_stable': impressions_stable,
                    'current_impressions': current_impressions
                }
            )
            
            return issue
        
        return None
    
    def _detect_impression_anomaly(
        self,
        current_data: Dict,
        historical_data: Dict,
        metrics: SEOMetrics
    ) -> Optional[AuditIssue]:
        """Detect significant impression drops"""
        
        current_impressions = metrics.total_impressions
        previous_impressions = historical_data.get('previous_impressions', 0)
        
        if previous_impressions == 0:
            return None
        
        # Calculate percentage change
        change_pct = ((current_impressions - previous_impressions) / previous_impressions) * 100
        
        # Determine severity
        severity = None
        thresholds = self.change_thresholds['impressions']
        
        if change_pct <= thresholds['critical']:
            severity = IssueSeverity.CRITICAL
        elif change_pct <= thresholds['high']:
            severity = IssueSeverity.HIGH
        elif change_pct <= thresholds['medium']:
            severity = IssueSeverity.MEDIUM
        
        if severity:
            issue = AuditIssue(
                issue_type=IssueType.IMPRESSION_DROP,
                severity=severity,
                category=IssueCategory.PERFORMANCE,
                title="Search Visibility Declined",
                description=f"Search impressions have dropped by {abs(change_pct):.1f}%. "
                           f"Current: {current_impressions:,} impressions, Previous: {previous_impressions:,} impressions.",
                recommendation="Check for indexing issues, review your sitemap, "
                              "and ensure important pages are not blocked by robots.txt.",
                traffic_impact=abs(change_pct),
                business_impact="high",
                evidence={
                    'current_impressions': current_impressions,
                    'previous_impressions': previous_impressions,
                    'change_percentage': change_pct
                }
            )
            
            return issue
        
        return None
    
    def _detect_statistical_anomalies(
        self,
        current_data: Dict,
        historical_data: Dict,
        metrics: SEOMetrics
    ) -> List[AuditIssue]:
        """Detect anomalies using statistical methods (Z-score)"""
        
        issues = []
        
        # Check clicks anomaly
        if 'clicks_mean' in historical_data and 'clicks_stdev' in historical_data:
            if historical_data['clicks_stdev'] > 0:
                z_score = abs((metrics.total_clicks - historical_data['clicks_mean']) / historical_data['clicks_stdev'])
                
                if z_score >= self.z_score_thresholds['critical']:
                    issues.append(self._create_statistical_anomaly_issue(
                        'clicks', metrics.total_clicks, historical_data['clicks_mean'],
                        z_score, IssueSeverity.HIGH
                    ))
                elif z_score >= self.z_score_thresholds['warning']:
                    issues.append(self._create_statistical_anomaly_issue(
                        'clicks', metrics.total_clicks, historical_data['clicks_mean'],
                        z_score, IssueSeverity.MEDIUM
                    ))
        
        # Check impressions anomaly
        if 'impressions_mean' in historical_data and 'impressions_stdev' in historical_data:
            if historical_data['impressions_stdev'] > 0:
                z_score = abs((metrics.total_impressions - historical_data['impressions_mean']) / 
                             historical_data['impressions_stdev'])
                
                if z_score >= self.z_score_thresholds['critical']:
                    issues.append(self._create_statistical_anomaly_issue(
                        'impressions', metrics.total_impressions, historical_data['impressions_mean'],
                        z_score, IssueSeverity.HIGH
                    ))
        
        return issues
    
    def _create_statistical_anomaly_issue(
        self,
        metric_name: str,
        current_value: float,
        expected_value: float,
        z_score: float,
        severity: IssueSeverity
    ) -> AuditIssue:
        """Create an issue for statistical anomaly"""
        
        deviation_pct = abs((current_value - expected_value) / expected_value * 100) if expected_value != 0 else 0
        
        return AuditIssue(
            issue_type=IssueType.TRAFFIC_DROP if metric_name == 'clicks' else IssueType.IMPRESSION_DROP,
            severity=severity,
            category=IssueCategory.PERFORMANCE,
            title=f"Statistical Anomaly in {metric_name.capitalize()}",
            description=f"The {metric_name} metric shows a significant statistical deviation. "
                       f"Current: {current_value:.0f}, Expected: {expected_value:.0f} "
                       f"(Z-score: {z_score:.2f}, {deviation_pct:.1f}% deviation).",
            recommendation=f"This unusual pattern may indicate a technical issue or external factor. "
                          f"Review recent changes and check for any errors or penalties.",
            traffic_impact=deviation_pct,
            business_impact="medium",
            evidence={
                'metric': metric_name,
                'current_value': current_value,
                'expected_value': expected_value,
                'z_score': z_score,
                'deviation_percentage': deviation_pct,
                'confidence_level': '99.7%' if z_score >= 3 else '95%'
            }
        )

File: api/internal/modules/audit/analyzers/anomaly.go
package analyzers

import (
	"fmt"
	"math"

	"github.com/petpeevephobia/solvia-v2/api/internal/infrastructure/google"
	"github.com/petpeevephobia/solvia-v2/api/internal/modules/audit/domain"
)

// ChangeThreshold defines percentage thresholds for different severity levels (1:1 with Python)
type ChangeThreshold struct {
	Critical float64
	High     float64
	Medium   float64
}

// AnomalyDetector detects statistical anomalies in GSC data
// Implements Z-score based anomaly detection (1:1 with Python audit_engine.py)
type AnomalyDetector struct {
	// Z-score thresholds for anomaly detection
	zScoreThresholdHigh   float64 // Above this = severe anomaly (3.0 = 99.7% confidence)
	zScoreThresholdMedium float64 // Above this = moderate anomaly (2.0 = 95% confidence)

	// Percentage change thresholds (1:1 with Python change_thresholds)
	trafficThresholds    ChangeThreshold
	positionThresholds   ChangeThreshold
	ctrThresholds        ChangeThreshold
	impressionThresholds ChangeThreshold

	// Minimum data points required for reliable statistics
	minDataPoints int

	// Minimum impressions to consider meaningful
	minImpressions int
}

// NewAnomalyDetector creates a new anomaly detector with default thresholds (1:1 with Python)
func NewAnomalyDetector() *AnomalyDetector {
	return &AnomalyDetector{
		// Z-score thresholds (1:1 with Python z_score_thresholds)
		zScoreThresholdHigh:   3.0, // 3.0 standard deviations = critical (99.7% confidence)
		zScoreThresholdMedium: 2.0, // 2.0 standard deviations = warning (95% confidence)

		// Percentage change thresholds (1:1 with Python change_thresholds)
		trafficThresholds: ChangeThreshold{
			Critical: -50, // >50% traffic loss
			High:     -20, // 20-50% traffic loss
			Medium:   -10, // 10-20% traffic loss
		},
		positionThresholds: ChangeThreshold{
			Critical: 5, // Dropped 5+ positions
			High:     3, // Dropped 3-5 positions
			Medium:   2, // Dropped 2-3 positions
		},
		ctrThresholds: ChangeThreshold{
			Critical: -50, // >50% CTR drop
			High:     -30, // 30-50% CTR drop
			Medium:   -15, // 15-30% CTR drop
		},
		impressionThresholds: ChangeThreshold{
			Critical: -60, // >60% impression loss
			High:     -30, // 30-60% impression loss
			Medium:   -15, // 15-30% impression loss
		},

		minDataPoints:  7,  // Need at least a week of data
		minImpressions: 50, // Minimum impressions to analyze
	}
}

// AnomalyResult contains detected anomalies
type AnomalyResult struct {
	Issues        []domain.AuditIssue
	Anomalies     []Anomaly
	HasAnomalies  bool
}

// Anomaly represents a detected statistical anomaly
type Anomaly struct {
	Type        string  `json:"type"`        // "drop", "spike", "volatility"
	Metric      string  `json:"metric"`      // "impressions", "clicks", "ctr", "position"
	Date        string  `json:"date"`
	Value       float64 `json:"value"`
	ExpectedMin float64 `json:"expected_min"`
	ExpectedMax float64 `json:"expected_max"`
	ZScore      float64 `json:"z_score"`
	Severity    string  `json:"severity"`    // "high", "medium", "low"
}

// Analyze performs anomaly detection on time series data
func (d *AnomalyDetector) Analyze(dailyMetrics []google.DailyMetric, auditID int64) *AnomalyResult {
	result := &AnomalyResult{
		Issues:       []domain.AuditIssue{},
		Anomalies:    []Anomaly{},
		HasAnomalies: false,
	}

	// Need sufficient data for statistical analysis
	if len(dailyMetrics) < d.minDataPoints {
		return result
	}

	// Detect anomalies in each metric using Z-score (statistical method)
	impressionAnomalies := d.detectAnomaliesInSeries(extractImpressions(dailyMetrics), dailyMetrics, "impressions")
	clickAnomalies := d.detectAnomaliesInSeries(extractClicks(dailyMetrics), dailyMetrics, "clicks")
	ctrAnomalies := d.detectAnomaliesInSeries(extractCTR(dailyMetrics), dailyMetrics, "ctr")
	positionAnomalies := d.detectAnomaliesInSeries(extractPosition(dailyMetrics), dailyMetrics, "position")

	// Combine all anomalies
	result.Anomalies = append(result.Anomalies, impressionAnomalies...)
	result.Anomalies = append(result.Anomalies, clickAnomalies...)
	result.Anomalies = append(result.Anomalies, ctrAnomalies...)
	result.Anomalies = append(result.Anomalies, positionAnomalies...)

	result.HasAnomalies = len(result.Anomalies) > 0

	// Convert significant anomalies to issues
	result.Issues = d.convertAnomaliestoIssues(result.Anomalies, auditID)

	return result
}

// AnalyzeWithComparison performs percentage-based anomaly detection comparing current vs previous period
// This is 1:1 with Python's detect_anomalies() method
func (d *AnomalyDetector) AnalyzeWithComparison(
	currentClicks, previousClicks int,
	currentImpressions, previousImpressions int,
	currentCTR, previousCTR float64,
	currentPosition, previousPosition float64,
	auditID int64,
) []domain.AuditIssue {
	var issues []domain.AuditIssue

	// 1. Traffic Drop Detection (1:1 with Python _detect_traffic_anomaly)
	if issue := d.detectTrafficPercentageAnomaly(currentClicks, previousClicks, auditID); issue != nil {
		issues = append(issues, *issue)
	}

	// 2. Position Loss Detection (1:1 with Python _detect_position_anomaly)
	if issue := d.detectPositionPercentageAnomaly(currentPosition, previousPosition, auditID); issue != nil {
		issues = append(issues, *issue)
	}

	// 3. CTR Decline Detection (1:1 with Python _detect_ctr_anomaly)
	if issue := d.detectCTRPercentageAnomaly(currentCTR, previousCTR, currentImpressions, previousImpressions, auditID); issue != nil {
		issues = append(issues, *issue)
	}

	// 4. Impression Drop Detection (1:1 with Python _detect_impression_anomaly)
	if issue := d.detectImpressionPercentageAnomaly(currentImpressions, previousImpressions, auditID); issue != nil {
		issues = append(issues, *issue)
	}

	return issues
}

// detectTrafficPercentageAnomaly detects significant traffic drops (1:1 with Python _detect_traffic_anomaly)
func (d *AnomalyDetector) detectTrafficPercentageAnomaly(currentClicks, previousClicks int, auditID int64) *domain.AuditIssue {
	if previousClicks == 0 {
		return nil
	}

	changePct := float64(currentClicks-previousClicks) / float64(previousClicks) * 100

	var severity string
	if changePct <= d.trafficThresholds.Critical {
		severity = "critical"
	} else if changePct <= d.trafficThresholds.High {
		severity = "high"
	} else if changePct <= d.trafficThresholds.Medium {
		severity = "medium"
	} else {
		return nil
	}

	return &domain.AuditIssue{
		AuditID:  auditID,
		Severity: severity,
		Category: "anomaly",
		Title:    "Significant Traffic Drop Detected",
		Description: fmt.Sprintf(
			"Organic traffic has dropped by %.1f%% compared to the previous period. Current: %d clicks, Previous: %d clicks.",
			math.Abs(changePct), currentClicks, previousClicks,
		),
		Impact:         "Fewer clicks means less organic traffic to your website, potentially impacting leads and conversions.",
		Recommendation: "Investigate recent changes to your site, check for algorithm updates, and review your top-performing pages for issues.",
	}
}

// detectPositionPercentageAnomaly detects significant position losses (1:1 with Python _detect_position_anomaly)
func (d *AnomalyDetector) detectPositionPercentageAnomaly(currentPosition, previousPosition float64, auditID int64) *domain.AuditIssue {
	if previousPosition == 0 || currentPosition == 0 {
		return nil
	}

	// Position change (positive means dropped in rankings - higher position number is worse)
	positionChange := currentPosition - previousPosition

	var severity string
	if positionChange >= d.positionThresholds.Critical {
		severity = "critical"
	} else if positionChange >= d.positionThresholds.High {
		severity = "high"
	} else if positionChange >= d.positionThresholds.Medium {
		severity = "medium"
	} else {
		return nil
	}

	return &domain.AuditIssue{
		AuditID:  auditID,
		Severity: severity,
		Category: "anomaly",
		Title:    "Average Position Declined",
		Description: fmt.Sprintf(
			"Your average search position has dropped from %.1f to %.1f (declined by %.1f positions).",
			previousPosition, currentPosition, positionChange,
		),
		Impact:         "Worse rankings mean your pages appear lower in search results, reducing visibility and clicks.",
		Recommendation: "Review your content quality, check for new competitors, and ensure your pages are optimized for target keywords.",
	}
}

// detectCTRPercentageAnomaly detects CTR problems (1:1 with Python _detect_ctr_anomaly)
func (d *AnomalyDetector) detectCTRPercentageAnomaly(currentCTR, previousCTR float64, currentImpressions, previousImpressions int, auditID int64) *domain.AuditIssue {
	if previousCTR == 0 {
		return nil
	}

	ctrChangePct := (currentCTR - previousCTR) / previousCTR * 100

	// Special case: CTR declined but impressions are stable/growing
	impressionsStable := float64(currentImpressions) >= float64(previousImpressions)*0.9

	var severity string
	if ctrChangePct <= d.ctrThresholds.Critical {
		if impressionsStable {
			severity = "critical"
		} else {
			severity = "high"
		}
	} else if ctrChangePct <= d.ctrThresholds.High {
		if impressionsStable {
			severity = "high"
		} else {
			severity = "medium"
		}
	} else if ctrChangePct <= d.ctrThresholds.Medium && impressionsStable {
		severity = "medium"
	} else {
		return nil
	}

	title := "Click-Through Rate Declined"
	description := fmt.Sprintf(
		"CTR has dropped from %.2f%% to %.2f%% (%.1f%% decline).",
		previousCTR*100, currentCTR*100, math.Abs(ctrChangePct),
	)
	if impressionsStable {
		title = "Click-Through Rate Declined Despite Stable Impressions"
		description += " Your pages are showing in search but not attracting clicks."
	}

	return &domain.AuditIssue{
		AuditID:        auditID,
		Severity:       severity,
		Category:       "anomaly",
		Title:          title,
		Description:    description,
		Impact:         "Lower CTR means fewer visitors despite search visibility.",
		Recommendation: "Review and improve your title tags and meta descriptions. Ensure they are compelling and match search intent.",
	}
}

// detectImpressionPercentageAnomaly detects significant impression drops (1:1 with Python _detect_impression_anomaly)
func (d *AnomalyDetector) detectImpressionPercentageAnomaly(currentImpressions, previousImpressions int, auditID int64) *domain.AuditIssue {
	if previousImpressions == 0 {
		return nil
	}

	changePct := float64(currentImpressions-previousImpressions) / float64(previousImpressions) * 100

	var severity string
	if changePct <= d.impressionThresholds.Critical {
		severity = "critical"
	} else if changePct <= d.impressionThresholds.High {
		severity = "high"
	} else if changePct <= d.impressionThresholds.Medium {
		severity = "medium"
	} else {
		return nil
	}

	return &domain.AuditIssue{
		AuditID:  auditID,
		Severity: severity,
		Category: "anomaly",
		Title:    "Search Visibility Declined",
		Description: fmt.Sprintf(
			"Search impressions have dropped by %.1f%%. Current: %d impressions, Previous: %d impressions.",
			math.Abs(changePct), currentImpressions, previousImpressions,
		),
		Impact:         "Declining impressions indicate your site is being shown less in search results.",
		Recommendation: "Check for indexing issues, review your sitemap, and ensure important pages are not blocked by robots.txt.",
	}
}

// detectAnomaliesInSeries detects anomalies in a single metric series
func (d *AnomalyDetector) detectAnomaliesInSeries(values []float64, dailyMetrics []google.DailyMetric, metricName string) []Anomaly {
	var anomalies []Anomaly

	if len(values) < d.minDataPoints {
		return anomalies
	}

	// Calculate mean and standard deviation
	mean := calculateMean(values)
	stdDev := calculateStdDev(values, mean)

	// Skip if no variation (all same values)
	if stdDev == 0 {
		return anomalies
	}

	// Check each value for anomalies
	for i, value := range values {
		zScore := (value - mean) / stdDev

		// Detect drops (significant negative z-score)
		if zScore < -d.zScoreThresholdHigh {
			anomalies = append(anomalies, Anomaly{
				Type:        "drop",
				Metric:      metricName,
				Date:        dailyMetrics[i].Date,
				Value:       value,
				ExpectedMin: mean - (2 * stdDev),
				ExpectedMax: mean + (2 * stdDev),
				ZScore:      zScore,
				Severity:    "high",
			})
		} else if zScore < -d.zScoreThresholdMedium {
			anomalies = append(anomalies, Anomaly{
				Type:        "drop",
				Metric:      metricName,
				Date:        dailyMetrics[i].Date,
				Value:       value,
				ExpectedMin: mean - (2 * stdDev),
				ExpectedMax: mean + (2 * stdDev),
				ZScore:      zScore,
				Severity:    "medium",
			})
		}

		// Detect spikes (significant positive z-score)
		// Note: For position, a spike (higher value) is actually bad
		if zScore > d.zScoreThresholdHigh {
			anomalyType := "spike"
			severity := "medium" // Spikes are often less critical than drops

			// For position, a spike means worse rankings - this is bad
			if metricName == "position" {
				severity = "high"
			}

			anomalies = append(anomalies, Anomaly{
				Type:        anomalyType,
				Metric:      metricName,
				Date:        dailyMetrics[i].Date,
				Value:       value,
				ExpectedMin: mean - (2 * stdDev),
				ExpectedMax: mean + (2 * stdDev),
				ZScore:      zScore,
				Severity:    severity,
			})
		}
	}

	// Detect volatility (high coefficient of variation)
	cv := (stdDev / mean) * 100
	if cv > 50 && metricName == "impressions" {
		// High volatility in impressions might indicate ranking instability
		anomalies = append(anomalies, Anomaly{
			Type:        "volatility",
			Metric:      metricName,
			Date:        "",
			Value:       cv,
			ExpectedMin: 0,
			ExpectedMax: 30, // CV above 30% is high
			ZScore:      0,
			Severity:    "medium",
		})
	}

	return anomalies
}

// convertAnomaliestoIssues converts detected anomalies to audit issues
func (d *AnomalyDetector) convertAnomaliestoIssues(anomalies []Anomaly, auditID int64) []domain.AuditIssue {
	var issues []domain.AuditIssue

	// Group anomalies by type to avoid duplicate issues
	hasTrafficDrop := false
	hasRankingSpike := false
	hasVolatility := false

	for _, anomaly := range anomalies {
		switch {
		case anomaly.Type == "drop" && anomaly.Metric == "impressions" && !hasTrafficDrop && anomaly.Severity == "high":
			hasTrafficDrop = true
			issues = append(issues, domain.AuditIssue{
				AuditID:     auditID,
				Severity:    "high",
				Category:    "anomaly",
				Title:       "Unusual Traffic Drop Detected",
				Description: fmt.Sprintf("On %s, impressions dropped to %.0f, which is %.1f standard deviations below your average.",
					anomaly.Date, anomaly.Value, math.Abs(anomaly.ZScore)),
				Impact:      "Sudden traffic drops may indicate algorithm updates, technical issues, or manual actions.",
				Recommendation:  "Check Google Search Console for manual actions, review recent site changes, and monitor competitor activity.",
			})

		case anomaly.Type == "spike" && anomaly.Metric == "position" && !hasRankingSpike && anomaly.Severity == "high":
			hasRankingSpike = true
			issues = append(issues, domain.AuditIssue{
				AuditID:     auditID,
				Severity:    "high",
				Category:    "anomaly",
				Title:       "Sudden Ranking Drop Detected",
				Description: fmt.Sprintf("On %s, average position worsened to %.1f, significantly below your normal range.",
					anomaly.Date, anomaly.Value),
				Impact:      "Sudden ranking drops can dramatically reduce organic traffic and clicks.",
				Recommendation:  "Review recent content changes, check for technical SEO issues, and analyze competitor activity.",
			})

		case anomaly.Type == "volatility" && !hasVolatility:
			hasVolatility = true
			issues = append(issues, domain.AuditIssue{
				AuditID:     auditID,
				Severity:    "medium",
				Category:    "anomaly",
				Title:       "High Traffic Volatility",
				Description: fmt.Sprintf("Your %s show high variability (CV: %.1f%%), indicating unstable search performance.",
					anomaly.Metric, anomaly.Value),
				Impact:      "Volatile traffic makes it difficult to predict and plan for organic growth.",
				Recommendation:  "Focus on building a diverse keyword portfolio and consistent content publishing schedule.",
			})
		}
	}

	return issues
}

// Helper functions for extracting metric values from daily data

func extractImpressions(dailyMetrics []google.DailyMetric) []float64 {
	values := make([]float64, len(dailyMetrics))
	for i, m := range dailyMetrics {
		values[i] = float64(m.Impressions)
	}
	return values
}

func extractClicks(dailyMetrics []google.DailyMetric) []float64 {
	values := make([]float64, len(dailyMetrics))
	for i, m := range dailyMetrics {
		values[i] = float64(m.Clicks)
	}
	return values
}

func extractCTR(dailyMetrics []google.DailyMetric) []float64 {
	values := make([]float64, len(dailyMetrics))
	for i, m := range dailyMetrics {
		values[i] = m.CTR
	}
	return values
}

func extractPosition(dailyMetrics []google.DailyMetric) []float64 {
	values := make([]float64, len(dailyMetrics))
	for i, m := range dailyMetrics {
		values[i] = m.Position
	}
	return values
}

// Statistical helper functions

func calculateMean(values []float64) float64 {
	if len(values) == 0 {
		return 0
	}
	sum := 0.0
	for _, v := range values {
		sum += v
	}
	return sum / float64(len(values))
}

func calculateStdDev(values []float64, mean float64) float64 {
	if len(values) < 2 {
		return 0
	}
	sumSquares := 0.0
	for _, v := range values {
		diff := v - mean
		sumSquares += diff * diff
	}
	variance := sumSquares / float64(len(values)-1)
	return math.Sqrt(variance)
}
